File: boat-common/src/test/java/test/java/xyz/srclab/common/reflect/TypeRefTest.java
package test.java.xyz.srclab.common.reflect;

import org.testng.Assert;
import org.testng.annotations.Test;
import xyz.srclab.common.reflect.TypeRef;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.lang.reflect.WildcardType;
import java.util.List;
import java.util.Map;

/**
 * @author sunqian
 */
public class TypeRefTest {

    @Test
    public void testTypeRef() {
        TypeRef<Map<String, List<? extends String>>> typeRef = new TypeRef<Map<String, List<? extends String>>>() {
        };
        Type type = typeRef.type();
        Assert.assertTrue(type instanceof ParameterizedType);
        ParameterizedType mapType = (ParameterizedType) type;
        Assert.assertEquals(mapType.getRawType(), Map.class);

        Type[] mapActualTypeArguments = mapType.getActualTypeArguments();
        Assert.assertEquals(mapActualTypeArguments[0], String.class);
        Assert.assertTrue(mapActualTypeArguments[1] instanceof ParameterizedType);
        ParameterizedType listType = (ParameterizedType) mapActualTypeArguments[1];
        Assert.assertEquals(listType.getRawType(), List.class);

        Type[] listActualTypeArguments = listType.getActualTypeArguments();
        Assert.assertTrue(listActualTypeArguments[0] instanceof WildcardType);
        WildcardType stringWildcardType = (WildcardType) listActualTypeArguments[0];
        Assert.assertEquals(stringWildcardType.getUpperBounds()[0], String.class);
    }
}

File: easy-projects/easy-common/src/test/kotlin/test/xyz/srclab/common/lang/TypeRefTest.kt
package test.xyz.srclab.common.lang

import org.testng.annotations.Test
import xyz.srclab.common.lang.ref.Ref
import xyz.srclab.common.reflect.type.TypeRef
import xyz.srclab.test.doAssertEquals

object TypeRefTest {

    private fun forGeneric(): Ref<String> {
        return Ref.ofEmpty()
    }

    @Test
    fun testTypeRef() {
        open class B : TypeRef<Ref<String>>()
        open class C : B()
        open class D : C()

        val generic = D().type
        doAssertEquals(
            generic,
            TypeRefTest::class.java.getDeclaredMethod("forGeneric").genericReturnType
        )
    }
}
